File: src/differential_equations_second_order_constant_hard.py
"""
differential_equations - second_order_constant (hard)
Generated: 2026-02-22T05:25:36.135861
"""

from problem_utils import *

def generate():
    problem_type = randint(1, 5)

    if problem_type == 1:
        # Repeated roots + IVP
        r = nonzero(-4, 4)
        b_coeff = -2 * r
        c_coeff = r ** 2
        y0 = nonzero(-5, 5)
        yp0 = nonzero(-6, 6)

        # y = (C1 + C2*t)*e^(r*t)
        # y(0) = C1 = y0
        C1_val = y0
        # y'(0) = r*C1 + C2 = yp0
        C2_val = yp0 - r * C1_val

        ans = (C1_val + C2_val * t) * exp(r * t)

        return problem(
            question=f"Solve the IVP: $y''{b_coeff:+d}y'{c_coeff:+d}y=0$, $y(0)={y0}$, $y'(0)={yp0}$",
            answer=ans,
            difficulty=(1600, 1720),
            topic="differential_equations/second_order_constant",
            solution=steps(
                f"Characteristic equation: $(r-{r})^2=0$; repeated root $r={r}$",
                f"General solution: $y=(C_1+C_2 t)e^{{{r}t}}$",
                f"$y(0)={y0}$: $C_1={y0}$",
                f"$y'(0)={yp0}$: ${r}\\cdot{y0}+C_2={yp0}$, so $C_2={C2_val}$",
                f"$y={latex(ans)}$"
            ),
            calculator="scientific"
        )

    elif problem_type == 2:
        # Complex roots + IVP
        alpha = nonzero(-4, 4)
        beta = randint(2, 6)
        b_coeff = -2 * alpha
        c_coeff = alpha ** 2 + beta ** 2
        y0 = nonzero(-5, 5)
        yp0 = nonzero(-6, 6)

        C1_val = y0
        C2_val = Rational(yp0 - alpha * C1_val, beta)

        ans = exp(alpha * t) * (C1_val * cos(beta * t) + C2_val * sin(beta * t))

        return problem(
            question=f"Solve the IVP: $y''{b_coeff:+d}y'{c_coeff:+d}y=0$, $y(0)={y0}$, $y'(0)={yp0}$",
            answer=ans,
            difficulty=(1650, 1780),
            topic="differential_equations/second_order_constant",
            solution=steps(
                f"Roots: $r={alpha}\\pm{beta}i$",
                f"General: $y=e^{{{alpha}t}}(C_1\\cos({beta}t)+C_2\\sin({beta}t))$",
                f"$y(0)={y0}$: $C_1={y0}$",
                f"$y'(0)={yp0}$: ${alpha}\\cdot{y0}+{beta}C_2={yp0}$, $C_2={latex(C2_val)}$",
                f"$y={latex(ans)}$"
            ),
            calculator="scientific"
        )

    elif problem_type == 3:
        # Non-homogeneous y'' + b*y' + c*y = k*e^(a*t), a not a root
        r1 = nonzero(-4, 4)
        r2 = nonzero(-4, 4)
        while r2 == r1:
            r2 = nonzero(-4, 4)

        b_coeff = -(r1 + r2)
        c_coeff = r1 * r2

        a_val = nonzero(-4, 4)
        while a_val == r1 or a_val == r2:
            a_val = nonzero(-4, 4)
        k_val = nonzero(-5, 5)

        # y_p = k/(a^2 + b*a + c)*e^(a*t)
        denom = a_val**2 + b_coeff * a_val + c_coeff
        if denom == 0:
            a_val += 1
            denom = a_val**2 + b_coeff * a_val + c_coeff
        y_p = Rational(k_val, denom) * exp(a_val * t)

        C1, C2 = symbols('C_1 C_2')
        yh = C1 * exp(r1 * t) + C2 * exp(r2 * t)
        ans = yh + y_p

        return problem(
            question=f"Find the general solution: $y''{b_coeff:+d}y'{c_coeff:+d}y = {k_val}e^{{{a_val}t}}$",
            answer=ans,
            difficulty=(1700, 1820),
            topic="differential_equations/second_order_constant",
            solution=steps(
                f"Homogeneous solution: $y_h = C_1 e^{{{r1}t}} + C_2 e^{{{r2}t}}$",
                f"Try $y_p = Ae^{{{a_val}t}}$; substitute to get $({a_val**2}{b_coeff:+d}\\cdot{a_val}{c_coeff:+d})A = {k_val}$",
                f"$A = \\frac{{{k_val}}}{{{denom}}} = {latex(Rational(k_val, denom))}$",
                f"General solution: $y = {latex(ans)}$"
            ),
            calculator="scientific"
        )

    elif problem_type == 4:
        # Non-homogeneous y'' + b*y' + c*y = k*cos(omega*t) or k*sin(omega*t)
        r1 = nonzero(-4, 4)
        r2 = nonzero(-4, 4)
        while r2 == r1:
            r2 = nonzero(-4, 4)

        b_coeff = -(r1 + r2)
        c_coeff = r1 * r2

        omega = randint(1, 5)
        k_val = nonzero(-5, 5)
        trig_fn = choice(['cos', 'sin'])

        # For y_p = A*cos + B*sin when forcing is cos(omega*t):
        # (c - omega^2)*A + b*omega*B = k
        # -b*omega*A + (c - omega^2)*B = 0
        D = (c_coeff - omega**2)**2 + (b_coeff * omega)**2
        if D == 0:
            omega += 1
            D = (c_coeff - omega**2)**2 + (b_coeff * omega)**2

        if trig_fn == 'cos':
            A_val = Rational(k_val * (c_coeff - omega**2), D)
            B_val = Rational(-k_val * b_coeff * omega, D)
            y_p = A_val * cos(omega * t) + B_val * sin(omega * t)
            rhs_latex = f"{k_val}\\cos({omega}t)"
        else:
            A_val = Rational(k_val * b_coeff * omega, D)
            B_val = Rational(k_val * (c_coeff - omega**2), D)
            y_p = A_val * cos(omega * t) + B_val * sin(omega * t)
            rhs_latex = f"{k_val}\\sin({omega}t)"

        C1, C2 = symbols('C_1 C_2')
        yh = C1 * exp(r1 * t) + C2 * exp(r2 * t)
        ans = yh + y_p

        return problem(
            question=f"Find the general solution: $y''{b_coeff:+d}y'{c_coeff:+d}y = {rhs_latex}$",
            answer=ans,
            difficulty=(1750, 1880),
            topic="differential_equations/second_order_constant",
            solution=steps(
                f"Homogeneous: $y_h = C_1 e^{{{r1}t}} + C_2 e^{{{r2}t}}$",
                f"Try $y_p = A\\cos({omega}t) + B\\sin({omega}t)$",
                f"Substitute, equate coefficients: $A = {latex(A_val)}$, $B = {latex(B_val)}$",
                f"General solution: $y = {latex(ans)}$"
            ),
            calculator="scientific"
        )

    else:
        # Non-homogeneous with polynomial forcing: y'' + b*y' + c*y = poly
        r1 = nonzero(-3, 3)
        r2 = nonzero(-3, 3)
        while r2 == r1:
            r2 = nonzero(-3, 3)

        b_coeff = -(r1 + r2)
        c_coeff = r1 * r2

        n = randint(1, 2)
        coeffs_rhs = [nonzero(-4, 4) for _ in range(n + 1)]
        g_t = sum(coeffs_rhs[i] * t**i for i in range(n + 1))

        # Particular solution: y_p = sum of polynomial terms
        # For simplicity compute via undetermined coefficients
        # For degree n forcing: y_p is degree n polynomial (since c_coeff != 0)
        # y_p = a0 + a1*t + ... + an*t^n
        # Substitute into ODE and match coefficients
        poly_syms = symbols(f'a0:{n+1}')
        y_p_try = sum(poly_syms[i] * t**i for i in range(n + 1))
        eqn = diff(y_p_try, t, 2) + b_coeff * diff(y_p_try, t) + c_coeff * y_p_try - g_t
        eqn_expanded = expand(eqn)

        coeff_eqs = [eqn_expanded.coeff(t, i) for i in range(n + 1)]
        sol_dict = solve(coeff_eqs, poly_syms)
        y_p = y_p_try.subs(sol_dict)
        y_p = simplify(y_p)

        C1, C2 = symbols('C_1 C_2')
        yh = C1 * exp(r1 * t) + C2 * exp(r2 * t)
        ans = yh + y_p

        return problem(
            question=f"Find the general solution: $y''{b_coeff:+d}y'{c_coeff:+d}y = {latex(g_t)}$",
            answer=ans,
            difficulty=(1800, 1950),
            topic="differential_equations/second_order_constant",
            solution=steps(
                f"Homogeneous: $y_h = C_1 e^{{{r1}t}} + C_2 e^{{{r2}t}}$",
                f"Particular (undetermined coefficients): try $y_p = $ polynomial of degree {n}",
                f"Substituting and matching gives $y_p = {latex(y_p)}$",
                f"General solution: $y = {latex(ans)}$"
            ),
            calculator="scientific"
        )


emit(generate())

File: src/differential_equations_second_order_constant_hard.jl
# differential_equations - second_order_constant (hard)
# Generated: 2026-03-08T20:53:20.183845

include(joinpath(@__DIR__, "..", "julia", "src", "ProblemUtils.jl"))
using .ProblemUtils

@script x y begin
    set_topic!("differential_equations/second_order_constant")
    
    problem_type = choice(1:5)
    
    if problem_type == 1
        # Characteristic equation with distinct real roots
        r1 = randint(-8, -1)
        r2 = randint(1, 8)
        while r2 == -r1
            r2 = randint(1, 8)
        end
        
        b_coeff = -(r1 + r2)
        c_coeff = r1 * r2
        
        general_solution = "y = C_1 e^{$(r1)x} + C_2 e^{$(r2)x}"
        
        diff_eq = "y'' $(b_coeff >= 0 ? "+" : "")$(b_coeff)y' $(c_coeff >= 0 ? "+" : "")$(c_coeff)y = 0"
        
        problem(
            question="Solve the second-order linear homogeneous differential equation: \$$(diff_eq)\$",
            answer=general_solution,
            difficulty=(1800, 2000),
            answer_type="text",
            solution=steps(
                "Given: \$$(diff_eq)\$",
                "Characteristic equation: \$r^2 $(b_coeff >= 0 ? "+" : "")$(b_coeff)r $(c_coeff >= 0 ? "+" : "")$(c_coeff) = 0\$",
                "Factor: \$(r - $(r1))(r - $(r2)) = 0\$",
                "Roots: \$r_1 = $(r1), r_2 = $(r2)\$ (distinct real roots)",
                sol("General solution", general_solution)
            ),
            time=180
        )
        
    elseif problem_type == 2
        # Characteristic equation with repeated real root
        r = randint(-6, 6)
        while r == 0
            r = randint(-6, 6)
        end
        
        b_coeff = -2*r
        c_coeff = r^2
        
        general_solution = "y = C_1 e^{$(r)x} + C_2 x e^{$(r)x}"
        
        diff_eq = "y'' $(b_coeff >= 0 ? "+" : "")$(b_coeff)y' $(c_coeff >= 0 ? "+" : "")$(c_coeff)y = 0"
        
        problem(
            question="Solve the second-order linear homogeneous differential equation: \$$(diff_eq)\$",
            answer=general_solution,
            difficulty=(1900, 2100),
            answer_type="text",
            solution=steps(
                "Given: \$$(diff_eq)\$",
                "Characteristic equation: \$r^2 $(b_coeff >= 0 ? "+" : "")$(b_coeff)r $(c_coeff >= 0 ? "+" : "")$(c_coeff) = 0\$",
                "Factor: \$(r - $(r))^2 = 0\$",
                "Repeated root: \$r = $(r)\$ (multiplicity 2)",
                sol("General solution", general_solution)
            ),
            time=200
        )
        
    elseif problem_type == 3
        # Complex conjugate roots
        alpha = randint(-5, 5)
        beta = randint(2, 8)
        
        b_coeff = -2*alpha
        c_coeff = alpha^2 + beta^2
        
        if alpha == 0
            general_solution = "y = C_1 \\cos($(beta)x) + C_2 \\sin($(beta)x)"
        else
            general_solution = "y = e^{$(alpha)x}(C_1 \\cos($(beta)x) + C_2 \\sin($(beta)x))"
        end
        
        diff_eq = "y'' $(b_coeff >= 0 ? "+" : "")$(b_coeff)y' $(c_coeff >= 0 ? "+" : "")$(c_coeff)y = 0"
        
        problem(
            question="Solve the second-order linear homogeneous differential equation: \$$(diff_eq)\$",
            answer=general_solution,
            difficulty=(2000, 2200),
            answer_type="text",
            solution=steps(
                "Given: \$$(diff_eq)\$",
                "Characteristic equation: \$r^2 $(b_coeff >= 0 ? "+" : "")$(b_coeff)r $(c_coeff >= 0 ? "+" : "")$(c_coeff) = 0\$",
                "Use quadratic formula: \$r = \\frac{$(- b_coeff) \\pm \\sqrt{$(b_coeff^2) - $(4*c_coeff)}}{2}\$",
                "Complex roots: \$r = $(alpha) \\pm $(beta)i\$",
                sol("General solution", general_solution)
            ),
            time=220
        )
        
    elseif problem_type == 4
        # IVP with distinct real roots
        r1 = randint(-6, -1)
        r2 = randint(1, 6)
        
        b_coeff = -(r1 + r2)
        c_coeff = r1 * r2
        
        y0 = randint(-8, 8)
        yp0 = randint(-15, 15)
        
        C1 = (yp0 - r2*y0) // (r1 - r2)
        C2 = (yp0 - r1*y0) // (r2 - r1)
        
        if C1 == 0 && C2 == 0
            C1 = 1
            C2 = 1
            y0 = C1 + C2
            yp0 = r1*C1 + r2*C2
        end
        
        particular_solution = "y = "
        if C1 != 0
            particular_solution *= "$(C1)e^{$(r1)x}"
        end
        if C2 != 0
            if C1 != 0
                particular_solution *= C2 > 0 ? " + " : " - "
                particular_solution *= "$(abs(C2))e^{$(r2)x}"
            else
                particular_solution *= "$(C2)e^{$(r2)x}"
            end
        end
        
        diff_eq = "y'' $(b_coeff >= 0 ? "+" : "")$(b_coeff)y' $(c_coeff >= 0 ? "+" : "")$(c_coeff)y = 0"
        
        problem(
            question="Solve the initial value problem: \$$(diff_eq)\$, \$y(0) = $(y0)\$, \$y'(0) = $(yp0)\$",
            answer=particular_solution,
            difficulty=(2100, 2300),
            answer_type="text",
            solution=steps(
                "Characteristic equation: \$r^2 $(b_coeff >= 0 ? "+" : "")$(b_coeff)r $(c_coeff >= 0 ? "+" : "")$(c_coeff) = 0\$",
                "Roots: \$r_1 = $(r1), r_2 = $(r2)\$",
                "General solution: \$y = C_1 e^{$(r1)x} + C_2 e^{$(r2)x}\$",
                "Apply initial conditions: \$y(0) = C_1 + C_2 = $(y0)\$, \$y'(0) = $(r1)C_1 + $(r2)C_2 = $(yp0)\$",
                "Solve system: \$C_1 = $(C1), C_2 = $(C2)\$",
                sol("Particular solution", particular_solution)
            ),
            time=240
        )
        
    else
        # IVP with complex roots
        alpha = randint(-3, 3)
        beta = randint(2, 6)
        
        b_coeff = -2*alpha
        c_coeff = alpha^2 + beta^2
        
        y0 = randint(-8, 8)
        yp0 = randint(-12, 12)
        
        C1 = y0
        C2 = (yp0 - alpha*y0) // beta
        
        if C1 == 0 && C2 == 0
            C1 = 1
            C2 = 1
            y0 = C1
            yp0 = alpha*C1 + beta*C2
        end
        
        if alpha == 0
            particular_solution = "y = $(C1)\\cos($(beta)x)"
            if C2 != 0
                particular_solution *= C2 > 0 ? " + " : " - "
                particular_solution *= "$(abs(C2))\\sin($(beta)x)"
            end
        else
            particular_solution = "y = e^{$(alpha)x}($(C1)\\cos($(beta)x)"
            if C2 != 0
                particular_solution *= C2 > 0 ? " + " : " - "
                particular_solution *= "$(abs(C2))\\sin($(beta)x))"
            else
                particular_solution *= ")"
            end
        end
        
        diff_eq = "y'' $(b_coeff >= 0 ? "+" : "")$(b_coeff)y' $(c_coeff >= 0 ? "+" : "")$(c_coeff)y = 0"
        
        problem(
            question="Solve the initial value problem: \$$(diff_eq)\$, \$y(0) = $(y0)\$, \$y'(0) = $(yp0)\$",
            answer=particular_solution,
            difficulty=(2200, 2400),
            answer_type="text",
            solution=steps(
                "Characteristic equation: \$r^2 $(b_coeff >= 0 ? "+" : "")$(b_coeff)r $(c_coeff >= 0 ? "+" : "")$(c_coeff) = 0\$",
                "Complex roots: \$r = $(alpha) \\pm $(beta)i\$",
                alpha == 0 ? "General solution: \$y = C_1\\cos($(beta)x) + C_2\\sin($(beta)x)\$" : "General solution: \$y = e^{$(alpha)x}(C_1\\cos($(beta)x) + C_2\\sin($(beta)x))\$",
                "Apply \$y(0) = $(y0)\$: \$C_1 = $(C1)\$",
                "Apply \$y'(0) = $(yp0)\$: \$C_2 = $(C2)\$",
                sol("Particular solution", particular_solution)
            ),
            time=260
        )
    end
end